zip-0317.rst

```diff
@@ -8,9 +8,308 @@
+  to reduce information leakage from input and output arity, and
+  is the standard approach used by `zcashd` and the mobile SDK
+  transaction builder.
+* Without a grace window, an input with value below the marginal
+  fee would never be worth including in the resulting transaction.
+  With a grace window, an input with value below :math:`marginal\_fee`
+  *is* worth including, if a second input is available that covers
+  both the primary output amount and the conventional transaction
+  fee.
+
+**Why a grace window of 2?**
+
+A 1-in, 2-out (or 2-action) transaction is the smallest possible
+transaction that permits both an output to a recipient, and a
+change output. However, as stated above, `zcashd` and the mobile
+SDK transaction builder will pad the number of inputs to at least 2.
+
+Let :math:`min\_actions` be the minimum number of logical actions
+that can be used to execute economically relevant transactions that
+produce change. Due to the aforementioned padding, :math:`min\_actions = 2`.
+
+Having a grace window size greater than :math:`min\_actions` would
+increase the cost to create such a minimal transaction. If the
+cost for a minimal transaction is bounded above by :math:`B`, then
+possible choices of :math:`marginal\_fee` are bounded above by
+:math:`B / max(min\_actions, grace\_actions)`. Therefore, the
+optimal choice of :math:`grace\_actions` to maximize the cost of
+denial-of-service attacks that use many logical actions, without
+imposing an undue penalty on minimal transactions, is
+:math:`grace\_actions = min\_actions = 2`.
+
+Marginal Fee
+~~~~~~~~~~~~
+
+This returns the conventional fee for a minimal transaction (as
+described above) to the original conventional fee of 10000 zatoshis
+specified in [#zip-0313]_, and imposes a non-trivial cost for
+potential denial-of-service attacks.
+
+Transaction relaying
+--------------------
+
+zcashd, zebrad, and potentially other node implementations, implement
+fee-based restrictions on relaying of mempool transactions. Nodes that
+normally relay transactions are expected to do so for transactions that pay
+at least the conventional fee as specified in this ZIP, unless there are
+other reasons not to do so for robustness or denial-of-service mitigation.
+
+Mempool size limiting
+---------------------
+
+zcashd and zebrad limit the size of the mempool as described in [#zip-0401]_.
+This specifies a :math:`low\_fee\_penalty` that is added to the "eviction weight"
+if the transaction pays a fee less than the conventional transaction fee.
+This threshold is modified to use the new conventional fee formula.
+
+Block production
+----------------
+
+Miners, mining pools, and other block producers, select transactions for
+inclusion in blocks using a variety of criteria. Where the criteria
+previously used the conventional transaction fee defined in ZIP 313 to
+decide on transaction inclusion, it is expected to instead use the formula
+specified in this ZIP.
```

Comment:
Add something about the incentives miners have to not fill blocks:
* they aren't currently getting significant fees;
* they get better orphan rates.